### PR TITLE
Use "batched" TRTRI & TRMM for Supernode-sptrsv setup

### PR DESCRIPTION
**Pull request by iyamazaki**
This PR integrates the "batched" TRTRI & TRTRM from [PR #697](https://github.com/kokkos/kokkos-kernels/pull/697) for sptrsv-supernode setup (it still runs on host, but now uses parallel-for for smaller blocks).

### CODE REVIEW

**Comment by ndellingwood** [comment]
@iyamazaki reports seeing this failure on white testbed, though the changes here are unrelated to the trsv routine and test and should not block this PR:

```bash
3: [ RUN      ] cuda.sparse_trsv_mv_kokkos_complex_double_int_int_LayoutLeft_TestExecSpace

3: /ascldap/users/iyamaza/github/tutorial/Kokkos/scratch/forks/sptrsv-supernode3/kokkos-kernels/unit_test/../test_common/KokkosKernels_TestUtils.hpp:88: Failure

3: The difference between double(AT1::abs(val1)) and double(AT2::abs(val2)) is 7.6400725210421072, which exceeds double(AT3::abs(tol)), where

3: double(AT1::abs(val1)) evaluates to 6.3380787715745086,

3: double(AT2::abs(val2)) evaluates to 13.978151292616616, and

3: double(AT3::abs(tol)) evaluates to 9.9999999999999995e-08.
```

@vqd8a would you be able to take a look at these failures at some point? On other machines this seemed to show up if the machine had extra workloads / resource contention, but on white there is a dedicated compute node.

**Comment by vqd8a** [comment]
@ndellingwood I will look at these.

File: src/sparse/KokkosSparse_sptrsv_supernode.hpp
Comment (e10harvey) [inline src/sparse/KokkosSparse_sptrsv_supernode.hpp:1197]: Consider moving these conditionals to https://github.com/kokkos/kokkos-kernels/pull/797/files#diff-3205ca442d81699d8344ea8328d91f81R1172.

File: src/sparse/KokkosSparse_sptrsv_supernode.hpp
Comment (e10harvey) [inline src/sparse/KokkosSparse_sptrsv_supernode.hpp:1241]: Same as above. Consider using a single `TriSupernodeTrtriFunctor` invocation by moving these conditionals to https://github.com/kokkos/kokkos-kernels/pull/797/files#diff-3205ca442d81699d8344ea8328d91f81R1172 and setting `using uplo_arg = Uplo::{Lower,Upper}` and `using diag_arg = Diag::{Unit,NonUnit}`.

File: src/sparse/KokkosSparse_sptrsv_supernode.hpp
Comment (e10harvey) [inline src/sparse/KokkosSparse_sptrsv_supernode.hpp:1188]: Consider experimenting with `nsuper` and determining a `parallel_switch`. If `nsuper` > `parallel_switch`, could we perform two passes:
1. parallel or serial scan `nb` for batched matrix and non-batched matrix indices.
2. apply block inversions in parallel on:
  a. batched matrices
  b. non-batched matrices
3. fence before reading the inversions

**Review by e10harvey — COMMENTED** [review]
@iyamazaki, Everything looks great. Please see a few considerations below. Are there any performance regressions with these changes?

Once spot-checks are passing, I will approve these changes.

File: src/sparse/KokkosSparse_sptrsv_supernode.hpp
Comment (iyamazaki) [reply src/sparse/KokkosSparse_sptrsv_supernode.hpp:1188]: Thank you for the review @e10harvey !! I'll try moving the if-statement out of functor.

In term of performance, if we run it with one-thread, it may run slower if batched TRTRI/TRTRM runs slower than device TRTRI/TRTRM for the small blocks on one core. We can play with the "size_unblock" which is used to decide when to switch between the batch vs. device calls. But this parameter is also used to decide when to switch during the solve. So, I may need to split the parameter to "size_unblock_setup" and "size_unblock_solve".

I am not sure if we can "apply non-batched/device calls in parallel"?

Comment (e10harvey) [reply src/sparse/KokkosSparse_sptrsv_supernode.hpp:1188]: > I am not sure if we can "apply non-batched/device calls in parallel"?

I am thinking about an effort to saturate the hardware; maybe two main host threads - one targeting the device(s) for non-batched matrices and another targeting all available host threads for batched matrices.

Comment (iyamazaki) [reply src/sparse/KokkosSparse_sptrsv_supernode.hpp:1188]: Wow.. You are thinking to split the threads for calling batched kernels, and for calling KokkosBlas::trtri & KokkosBlast::trmm in sequence (these may be threaded kernels, with TPL?), in parallel? Interesting idea, I can take a look (e.g., I am not sure how many threads to use on each, and big blocks may just dominate), maybe after this PR..

File: src/sparse/KokkosSparse_sptrsv_supernode.hpp
Comment (e10harvey) [reply src/sparse/KokkosSparse_sptrsv_supernode.hpp:1271]: @iyamazaki, @ndellingwood  To support a single `TriSupernodeTrtriFunctor` invocation the sptrsv handle and callers would need to be refactored such that [lower](https://github.com/kokkos/kokkos-kernels/blob/ad1c5d0ffd7b73bbc836d610a6acb96ddcc7cc88/src/sparse/KokkosSparse_sptrsv_handle.hpp#L247), [unit_diag](https://github.com/kokkos/kokkos-kernels/blob/ad1c5d0ffd7b73bbc836d610a6acb96ddcc7cc88/src/sparse/KokkosSparse_sptrsv_handle.hpp#L319), and [invert_off_diag](https://github.com/kokkos/kokkos-kernels/blob/ad1c5d0ffd7b73bbc836d610a6acb96ddcc7cc88/src/sparse/KokkosSparse_sptrsv_handle.hpp#L347) are template parameters. This seems like a significant effort but could improve readability and maintainability in the long run; is it worth creating a issue with a cleanup label for this?

Comment (ndellingwood) [reply src/sparse/KokkosSparse_sptrsv_supernode.hpp:1271]: @e10harvey good point about cleanup, would you be okay that we open a new issue and focus on that in a separate effort for this PR?

Comment (e10harvey) [reply src/sparse/KokkosSparse_sptrsv_supernode.hpp:1271]: @ndellingwood: Yes, I'll create an issue.

Comment (iyamazaki) [reply src/sparse/KokkosSparse_sptrsv_supernode.hpp:1271]: Thank you!! So, any additional changes are needed for this PR, or will be addressed by the next one, @e10harvey ?

Comment (e10harvey) [reply src/sparse/KokkosSparse_sptrsv_supernode.hpp:1271]: @iyamazaki I think this PR just needs spot-checks run. I'll run and post those later today.

**Comment by e10harvey** [comment]
## Spot-check
```
#######################################################
PASSED TESTS
#######################################################
clang-8.0-Cuda_OpenMP-release build_time=653 run_time=104
clang-8.0-Pthread_Serial-release build_time=208 run_time=130
clang-9.0.0-Pthread-release build_time=131 run_time=61
clang-9.0.0-Serial-release build_time=138 run_time=47
cuda-10.1-Cuda_OpenMP-release build_time=892 run_time=103
cuda-11.0-Cuda_OpenMP-release build_time=925 run_time=103
cuda-9.2-Cuda_Serial-release build_time=843 run_time=155
gcc-7.3.0-OpenMP-release build_time=196 run_time=50
gcc-7.3.0-Pthread-release build_time=130 run_time=70
gcc-8.3.0-Serial-release build_time=159 run_time=56
gcc-9.1-OpenMP-release build_time=222 run_time=52
gcc-9.1-Serial-release build_time=187 run_time=56
intel-17.0.1-Serial-release build_time=298 run_time=52
intel-18.0.5-OpenMP-release build_time=398 run_time=51
intel-19.0.5-Pthread-release build_time=401 run_time=63
```

### Spot-check-tpls
```
#######################################################
PASSED TESTS
#######################################################
cuda-10.1-Cuda_OpenMP-release build_time=922 run_time=106
cuda-11.0-Cuda_OpenMP-release build_time=958 run_time=134
intel-17.0.1-Serial-release build_time=317 run_time=53
intel-18.0.5-OpenMP-release build_time=378 run_time=48
intel-19.0.5-Pthread-release build_time=423 run_time=65
#######################################################
FAILED TESTS
#######################################################
clang-8.0-Cuda_OpenMP-release (build failed)
#######################################################
```

File: src/sparse/KokkosSparse_sptrsv_supernode.hpp
Comment (iyamazaki) [reply src/sparse/KokkosSparse_sptrsv_supernode.hpp:1271]: Thank you, @e10harvey !! Do you have some details for the failed build?

Comment (e10harvey) [reply src/sparse/KokkosSparse_sptrsv_supernode.hpp:1271]: > Thank you, @e10harvey !! Do you have some details for the failed build?

No problem. I do and I'm testing the fix now. If it passes, I'll post PR with that so we can get this merged.

**Review by e10harvey — APPROVED** [review]

**Pull request merged**
